entgql/extension.go

```diff
@@ -231,6 +285,299 @@ func (e *Extension) mapScalar(f *gen.Field, op gen.Op) string {
+		var (
+			queries    = make(map[string]*ast.FieldDefinition)
+			interfaces = make(map[string]*ast.InterfaceDefinition)
+			scalars    = make(map[string]*ast.ScalarDefinition)
+			objects    = make(map[string]*ast.ObjectDefinition)
+			enums      = make(map[string]*ast.EnumDefinition)
+			inputs     = make(map[string]*ast.InputObjectDefinition)
+		)
+		return gen.GenerateFunc(func(g *gen.Graph) error {
+			nodes, err := filterNodes(g.Nodes)
+			if err != nil {
+				return err
+			}
+			if err := next.Generate(g); err != nil {
+				return err
+			}
+
+			addTimeScalarDefinition(scalars)
+			addNodeDefinition(interfaces)
+			// Iterate through the nodes and create the base objects, enums and queries
+			for _, node := range nodes {
+				e.addEnumDefinitions(enums, node)
+				e.addObjectDefinition(objects, node)
+
+				var hasOrderByType bool
+				// If orderBy has been enabled, create the types for ordering
+				if e.orderBy {
+					hasOrderByType = e.addOrderByDefinitions(enums, inputs, node)
+				}
+
+				if e.whereFilters {
+					// Create the where input types
+					err := e.addWhereType(inputs, node)
+					if err != nil {
+						return err
+					}
+				}
+
+				if e.relaySpec {
+					e.addRelayObjectDefinitions(objects, node)
+					e.addQueryDefinition(queries, node, true, hasOrderByType)
+				}
+				// If either the relay spec is not enabled, or it is enabled but
+				// with a custom name, then generate also the "default" query
+				// for this node
+				if !e.relaySpec || e.relayQuery != nil {
+					e.addQueryDefinition(queries, node, false, hasOrderByType)
+				}
+			}
+			if e.orderBy {
+				// Create the global OrderDirection enum
+				e.addGlobalOrderByDefinition(enums)
+			}
+			if e.relaySpec {
+				// Create the global relay spec objects
+				addCursorScalarDefinition(scalars)
+				addPageInfoDefinition(objects)
+			}
+			return e.updateSchema(scalars, interfaces, queries, objects, enums, inputs)
+		})
+	}
+}
+
+// addObjectDefinition creates a GraphQL type for the given node, and adds it to the map of objects
+func (e *Extension) addObjectDefinition(objects map[string]*ast.ObjectDefinition, node *gen.Type) {
+	obj := ast.NewObjectDefinition(&ast.ObjectDefinition{
+		Name: newASTName(node.Name),
+		Description: ast.NewStringValue(&ast.StringValue{
+			Value: fmt.Sprintf("%s represents the node %s in the ent schema.\nGenerated by ent.", node.Name, node.Name),
+		}),
+		Fields: []*ast.FieldDefinition{
+			ast.NewFieldDefinition(&ast.FieldDefinition{
+				Name: newASTName("id"),
+				Type: ast.NewNonNull(&ast.NonNull{
+					Type: newASTNamed(graphql.ID.String()),
+				}),
+			}),
+		},
+		Interfaces: []*ast.Named{newASTNamed("Node")},
+	})
+
+	for _, f := range node.Fields {
+		obj.Fields = append(obj.Fields, ast.NewFieldDefinition(
+			&ast.FieldDefinition{
+				Name: newASTName(f.Name),
+				Type: newASTNamed(e.mapScalar(f, gen.EQ)),
+			}),
+		)
+	}
+
+	for _, edge := range node.Edges {
```

Comment:
we will need an annotation here. I don't think edges should be included by default.

Also some edges you will want to be connections and some arrays

Reply:
Agree. I had not thought about making an edge an array... and actually I can see some nice use cases for it !